assets/aml-benchmark/scripts/_internal/upgrade_components.py

```diff
@@ -0,0 +1,214 @@
+
+
+def _upgrade_component_env(spec: Dict[str, Any], spec_str: str) -> str:
+    """Upgrade component's environment."""
+    type = spec["type"]
+    if type == NodeType.COMMAND:
+        env_arr = spec["environment"].split("/")
+        if env_arr[-1] != "latest":
+            latest_version = _get_asset_latest_version(
+                asset_name=env_arr[-3],
+                asset_type=AzureMLResourceType.ENVIRONMENT,
+                asset_version=env_arr[-1],
+            )
+            env_arr[-1] = latest_version
+            spec_str = re.sub(
+                pattern=r"environment: .*",
+                repl=f"environment: {'/'.join(env_arr)}",
+                string=spec_str,
+            )
+    elif type == NodeType.PIPELINE:
+        spec_str = re.sub(
+            pattern=r"component: (\S+:\S+):(\d+\.\d+\.\d+)",
+            repl=__replace_pipeline_comp_job_version,
+            string=spec_str,
+        )
+    elif type == NodeType.PARALLEL:
+        env_arr = spec["task"]["environment"].split("/")
+        if env_arr[-1] != "latest":
+            latest_version = _get_asset_latest_version(
+                asset_name=env_arr[-3],
+                asset_type=AzureMLResourceType.ENVIRONMENT,
+                asset_version=env_arr[-1],
+            )
+            env_arr[-1] = latest_version
+            spec_str = re.sub(
+                pattern=r"environment: .*",
+                repl=f"environment: {'/'.join(env_arr)}",
+                string=spec_str,
+            )
+    return spec_str
+
+
+def _upgrade_component(
+    component_path: str,
+) -> Tuple[bool, Union[str, None], str, Optional[str]]:
+    """Upgrade component spec.
+
+    :param component_path: Path to component spec.
+    :return: Tuple of (error, error_message, component_path, component_name).
+    """
+    is_error = False
+    error = None
+    name = None
+    try:
+        with open(component_path, "r") as file:
+            spec = safe_load(file)
+            file.seek(0)
+            spec_str = file.read()
+        name = spec["name"]
+
+        # bump component version
+        latest_version = _get_asset_latest_version(
+            asset_name=name,
+            asset_type=AzureMLResourceType.COMPONENT,
+            asset_version=spec["version"],
+        )
+        new_version = _get_bumped_version(latest_version)
+        spec["version"] = new_version
+        spec_str = re.sub(
+            pattern=r"version: .*", repl=f"version: {new_version}", string=spec_str
+        )
+
+        # bump component's environment only where version is hardcoded
+        spec_str = _upgrade_component_env(spec, spec_str)
+
+        with open(component_path, "w") as file:
+            file.write(spec_str)
+    except Exception as e:
+        is_error = True
+        error = str(e)
+    return is_error, error, component_path, name
+
+
+def main():
+    """Entry function."""
+    components = _get_all_components_spec()
```

Comment:
Can we create flags to make upgrade environment, ignore resource_manager configurable? If possible, let us specify the environment version to pin or component version to pin to (instead of latest). Thanks!

Reply:
Added argument env_version` to control env upgrade (more details in readme.md). All the versions are now pinned to a specific version instead of `latest` label. Ignoring components for release is something that we should control within the script, there is a global variable for it which can be changed. Not adding this as an argument at the moment, does not seem to be worth doing at the moment.